### PR TITLE
ROC AUC implementation based on radix sort

### PR DESCRIPTION
**Pull request by gidler**
I implemented a Radix sort based ROC AUC function. My current calculations show this to be 5x-6x faster than the current C++ ROC AUC implementation. 

I replaced the existing C++ implementation in this PR so it would be easy to test, but we may consider putting it alongside the existing implementation. If we want to do that, I can update the PR accordingly.

WARNING: this method modifies the data behind the `float*` that is passed to it to avoid making a copy. If this is a problem, I can add a copy of the data or reset the data back to it's starting position, but this will add a additional runtime.

### CODE REVIEW

File: autogluon_zeroshot/metrics/_roc_auc_cpp/cpp_auc.cpp [hunk 1]
```diff
@@ -1,66 +1,131 @@
...
-        last_counted_tps = tps;
-    }
+inline uint32_t mantissa(uint32_t a) { return a & 0x7FFFFF; }
+
+/*
+ * This function is a faster single-threaded implementation of the binary non-weighted ROC AUC metric.
+ * The primary speed-up come from a radix sort on the 23 bits of the mantissa of float values.
+ * This function makes several assumptions:
+ *   - y_true and y_pred are both of size len
+ *   - y_pred values are in range [0.0, 1.0)
```
Comment (Innixma) [inline autogluon_zeroshot/metrics/_roc_auc_cpp/cpp_auc.cpp]: Why non-inclusive?

Comment (gidler) [reply autogluon_zeroshot/metrics/_roc_auc_cpp/cpp_auc.cpp]: It comes from the way floats are represented. But ultimately I want to sort numbers that come from a discrete range of floats that have a range of width 1, and [0.0, 1.0] is "technically" one bin larger than that. This discussion makes me think I need to add a corner case to make sure 1.0 is also handled correctly...

Comment (gidler) [reply autogluon_zeroshot/metrics/_roc_auc_cpp/cpp_auc.cpp]: Yes, additional testing shows this is indeed an issue. I will let you know when a fix is ready.

**Review by Innixma — APPROVED** [review]
Looks awesome! This is a 2x speedup from the prior C++ `roc_auc` implementation in the current simulation code!

Mainline:
```
1	: Train: 3.14 | 9.21s | ray | CatBoost_c1_BAG_L1
2	: Train: 2.57 | 44.74s | ray | NeuralNetFastAI_r44_BAG_L1
3	: Train: 2.34 | 61.77s | ray | LightGBM_r9_BAG_L1
4	: Train: 2.21 | 78.06s | ray | RandomForest_r19_BAG_L1
```
PR:
```
1	: Train: 3.14 | 8.77s | ray | CatBoost_c1_BAG_L1
2	: Train: 2.57 | 29.21s | ray | NeuralNetFastAI_r44_BAG_L1
3	: Train: 2.34 | 38.09s | ray | LightGBM_r9_BAG_L1
4	: Train: 2.21 | 46.21s | ray | RandomForest_r19_BAG_L1
```

**Pull request merged**